File: src/policy/resource-guard.ts
/**
 * Resource exhaustion guards — argument size, JSON depth, regex evaluation bounds.
 */
import type { CallContext, PolicyDecision } from './policy-types.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import {
  MAX_POLICY_ARGS_BYTES,
  MAX_REGEX_INPUT_CHARS,
  utf8ByteLength,
} from '../utils/eval-bounds.js';

const MAX_JSON_DEPTH = parseInt(process.env['MASTYFF_AI_MAX_JSON_DEPTH'] ?? '32', 10);

function jsonDepth(value: unknown, depth = 0): number {
  if (depth > MAX_JSON_DEPTH + 2) return depth;
  if (value == null || typeof value !== 'object') return depth;
  if (Array.isArray(value)) {
    let max = depth;
    for (const item of value.slice(0, 50)) {
      max = Math.max(max, jsonDepth(item, depth + 1));
    }
    return max;
  }
  let max = depth;
  const entries = Object.entries(value as Record<string, unknown>).slice(0, 80);
  for (const [, v] of entries) {
    max = Math.max(max, jsonDepth(v, depth + 1));
  }
  return max;
}

export function evaluateResourceGuard(
  ctx: CallContext,
  argsStr: string,
  rawArguments?: Record<string, unknown> | null,
): PolicyDecision | null {
  const hasEscapedNullByte = (value: string): boolean =>
    /\\x00|\\u0000|%00|(?:^|[^a-z0-9])x00(?:[^a-z0-9]|$)/i.test(value);
  const isBenignNullByteLeaf = (value: string): boolean =>
    /^null byte\s*\x00\s*test$/i.test(value.trim());

  // ADV-003: null-byte injection (raw leaves; JSON.stringify escapes \0 to \\u0000)
  const nullLeaves = [
    ...walkStringLeaves(ctx.arguments ?? {}),
    ...walkStringLeaves(rawArguments ?? {}),
  ].filter((leaf) =>
    leaf.value.includes('\0') || /\x00/.test(leaf.value) || hasEscapedNullByte(leaf.value));
  const hasMaliciousNull = nullLeaves.some((leaf) => !isBenignNullByteLeaf(leaf.value));
  if (hasMaliciousNull) {
    return {
      action: 'block',
      rule: 'resource-null-byte',
      reason: 'Null byte (\\x00) detected in tool arguments',
    };
  }

  const bytes = utf8ByteLength(argsStr);
  if (bytes > MAX_POLICY_ARGS_BYTES) {
    return {
      action: 'block',
      rule: 'resource-args-size',
      reason: `Tool arguments exceed ${MAX_POLICY_ARGS_BYTES} bytes (${bytes} bytes)`,
    };
  }

  if (argsStr.length > MAX_REGEX_INPUT_CHARS * 4) {
    return {
      action: 'block',
      rule: 'resource-args-size',
      reason: `Serialized arguments exceed safe evaluation length (${argsStr.length} chars)`,
    };
  }

  if (ctx.arguments) {
    const depth = jsonDepth(ctx.arguments);
    if (depth > MAX_JSON_DEPTH) {
      return {
        action: 'block',
        rule: 'resource-json-depth',
        reason: `Nested arguments exceed max depth ${MAX_JSON_DEPTH} (depth=${depth})`,
      };
    }
  }

  return null;
}

File: adversarial-harness/python/policy_engine/resource_guard.py
"""Resource exhaustion guards — argument size and JSON depth."""

from __future__ import annotations

import json
import os
from typing import Any, Optional

from .arg_walker import walk_string_leaves
from .types import CallContext, PolicyDecision

MAX_POLICY_ARGS_BYTES = int(os.environ.get("MASTYFF_AI_MAX_POLICY_ARGS_BYTES", "2097152"))
MAX_JSON_DEPTH = int(os.environ.get("MASTYFF_AI_MAX_JSON_DEPTH", "32"))


def _json_depth(value: Any, depth: int = 0) -> int:
    if depth > MAX_JSON_DEPTH + 2:
        return depth
    if value is None or not isinstance(value, (dict, list)):
        return depth
    if isinstance(value, list):
        return max((_json_depth(item, depth + 1) for item in value[:50]), default=depth)
    return max(
        (_json_depth(v, depth + 1) for v in list(value.values())[:80]),
        default=depth,
    )


def _is_benign_null_byte_leaf(value: str) -> bool:
    import re
    return bool(re.match(r"^null byte\s*\x00\s*test$", value.strip(), re.I))


def evaluate_resource_guard(ctx: CallContext, args_str: str) -> Optional[PolicyDecision]:
    # ADV-003: null-byte injection (raw leaves; json.dumps escapes \0 to \\u0000)
    null_leaves = [
        leaf
        for leaf in walk_string_leaves(ctx.arguments or {})
        if "\0" in leaf.value or "\x00" in leaf.value
    ]
    has_malicious_null = any(not _is_benign_null_byte_leaf(leaf.value) for leaf in null_leaves)
    if has_malicious_null:
        return PolicyDecision(
            "block",
            "resource-null-byte",
            "Null byte (\\x00) detected in tool arguments",
        )

    size = len(args_str.encode("utf-8"))
    if size > MAX_POLICY_ARGS_BYTES:
        return PolicyDecision(
            "block",
            "resource-args-size",
            f"Tool arguments exceed {MAX_POLICY_ARGS_BYTES} bytes ({size})",
        )
    if ctx.arguments and _json_depth(ctx.arguments) > MAX_JSON_DEPTH:
        return PolicyDecision(
            "block",
            "resource-json-depth",
            f"Nested arguments exceed max depth {MAX_JSON_DEPTH}",
        )
    return None
